mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/engine/rules/PawnRules.kt

```diff
@@ -0,0 +1,115 @@
+      when (color) {
+        Color.Black -> 7
+        Color.White -> 0
+      }
+
+  override fun AttackScope.attacks(color: Color, position: Position) {
+    val direction = direction(color)
+
+    attack(position + direction + CardinalPoints.E)
+    attack(position + direction + CardinalPoints.W)
+  }
+
+  /**
+   * Attempts to move or promote the pawn, depending on which position it's targeting.
+   *
+   * @param color the [Color] of the player.
+   * @param at the target [Position].
+   * @param effect the [Effect] to be applied.
+   */
+  private fun ActionScope.moveOrPromote(color: Color, at: Position, effect: Effect) {
+    if (endRow(color) == at.y) {
+      val choices = listOf(Bishop, Knight, Queen, Rook)
+      for (rank in choices) {
+        // For each possible rank, move the pawn, then remove it and finally replace it with an
+        // other piece with the same id.
+        promote(at, rank) {
+          effect()
+          val pawn = remove(at)
+          insert(at, MutableBoardPiece(PieceIdentifier(pawn.id), rank, color))
+        }
+      }
+    } else {
+      // Move without promotion.
+      move(at, effect)
+    }
+  }
+
+  /** Moves the pawn up, assuming there is a free square above it. */
+  private fun ActionScope.singleUp(color: Color, position: Position) {
+    val target = position + direction(color)
+    if (get(target).isNone) {
+      moveOrPromote(color = color, at = target) { move(from = position, to = target) }
+    }
+  }
+
+  /** Moves the pawn up by two squares, assuming that the two squares above are empty. */
+  private fun ActionScope.doubleUp(color: Color, position: Position) {
+    if (startRow(color) != position.y) return
+    val target = position + (direction(color) * 2)
+    if (get(position + direction(color)).isNone && get(target).isNone) {
+      move(at = target) { move(from = position, to = target) }
+    }
+  }
+
+  /** Takes the pieces to the right and to the left above the pawn. */
+  private fun ActionScope.sideTakes(color: Color, position: Position) {
+    for (sideDelta in listOf(CardinalPoints.E, CardinalPoints.W)) {
+      val target = position + direction(color) + sideDelta
+      val piece = get(target)
+      if (!piece.isNone && piece.color != color) {
+        moveOrPromote(color = color, at = target) { move(from = position, to = target) }
+      }
+    }
+  }
+
+  /** Takes the adversary pawns en-passant. */
+  private fun ActionScope.enPassant(color: Color, position: Position) {
+    for (sideDelta in listOf(CardinalPoints.E, CardinalPoints.W)) {
+      val target = position + direction(color) + sideDelta
+      val adversaryPosition = position + sideDelta
+      val adversary = get(adversaryPosition)
+      if (adversary.isNone || adversary.color == color || adversary.rank != Pawn) continue
+      val adversaryStart = target + direction(color)
+      if (!adversaryStart.inBounds) continue
+      if (getHistorical(adversaryStart).drop(2).any { it != adversary }) continue
```

Comment:
Question: I'm not sure I understand why you need to drop 2 boards from the history, instead of just 1. I don't think this would enable any wrong behavior since 2 boards ago would be an allied turn, but i'm still confused.

Reply:
Indeed, dropping one or two boards wouldn't change the behaviour. In the best case scenario, dropping two boards may avoid 1 extra comparison in the `any`, so the benefit of `drop(2)` over `drop(1)` is really marginal.

I'll add a comment that explains that either values could be used.

(also, I wish there was a `dropFirst()` method on `Sequence`, which could make this code clearer imho)

Reply:
Done in 6b261ff053d1e8037ec8b6e7bc05290a4ad188ae.

Reply:
Ah, I see! Very nice!